Type bounds on method not enforced (in presence of nullary method call)

## reproduction steps

using Scala 2.13.5

```scala
class Door[State <: DoorState] {
    def doorKey: Int = 1
    def open[Phantom >: State <: Closed.type]: Door[Open.type] = new Door[Open.type]
    def close[Phantom >: State <: Open.type]: Door[Closed.type] = new Door[Closed.type]
}

sealed trait DoorState
case object Closed extends DoorState
case object Open extends DoorState

val aClosedDoor = new Door[Closed.type]

println(aClosedDoor.open.close.close.doorKey == 1)  // How come this is compiling?
```

## problem

The compiler is not enforcing the type bounds that protect the access to method `close` (same thing happens for `open`) under this corner case.
Related SO question: https://stackoverflow.com/questions/67950913/type-bounds-on-methods-not-enforced

> because the implicit conversion converts to `Any`

I don't get this?  Why/where?  Converts to `Integer` no?

Btw, here's a comparison:

```
normalTypedApply(door.close[Phantom], List()) silentResult=SilentResultValue(door.close[Phantom])
normalTypedApply(door.close.toString, List()) silentResult=scala.tools.nsc.typechecker.Typers$SilentTypeError@44f75083
test/files/neg/t12413b.scala:10: error: inferred type arguments [AnyRef] do not conform to method close's type parameter bounds [Phantom >: AnyRef <: Open]
    door.close.toString()
         ^
1 error
```

To me this looks like it's failing 1 step too late: it should have already failed in the first one.

I don't understand why in the second test case (the minimised original one) it doesn't have that step until much later (second to last) - why does it end up fixing Int#toString first?

```
normalTypedApply(door.close.toString, List()) silentResult=SilentResultValue(door.close[Phantom].toString)
tryTypedApply(door.close[Phantom].toString, List()) res=scala.tools.nsc.typechecker.Typers$SilentTypeError@3d0f8e03
normalTypedApply(double2Double, List(<argument>)) silentResult=SilentResultValue(scala.Predef.double2Double)
tryTypedApply(scala.Predef.double2Double, List(<argument>)) res=SilentResultValue(scala.Predef.double2Double(<argument>.toDouble))
normalTypedApply(float2Float, List(<argument>)) silentResult=SilentResultValue(scala.Predef.float2Float)
tryTypedApply(scala.Predef.float2Float, List(<argument>)) res=SilentResultValue(scala.Predef.float2Float(<argument>.toFloat))
normalTypedApply(long2Long, List(<argument>)) silentResult=SilentResultValue(scala.Predef.long2Long)
tryTypedApply(scala.Predef.long2Long, List(<argument>)) res=SilentResultValue(scala.Predef.long2Long(<argument>.toLong))
normalTypedApply(int2Integer, List(<argument>)) silentResult=SilentResultValue(scala.Predef.int2Integer)
tryTypedApply(scala.Predef.int2Integer, List(<argument>)) res=SilentResultValue(scala.Predef.int2Integer(<argument>))
normalTypedApply(scala.Predef.int2Integer, List(door.close[Phantom])) silentResult=SilentResultValue(scala.Predef.int2Integer)
tryTypedApply(scala.Predef.int2Integer, List(door.close[Phantom])) res=SilentResultValue(scala.Predef.int2Integer(door.close[Phantom]))
adaptToArgs door.close[Phantom], name = toString, argtpes = List(), pt = Unit = scala.Predef.int2Integer(door.close[Phantom])
normalTypedApply(scala.Predef.int2Integer(door.close[Phantom]).toString, List()) silentResult=SilentResultValue(scala.Predef.int2Integer(door.close[Phantom]).toString)
normalTypedApply(Open.super.<init>, List()) silentResult=SilentResultValue(Open.super.<init>)
normalTypedApply(Door.super.<init>, List()) silentResult=SilentResultValue(Door.super.<init>)
normalTypedApply(Test.super.<init>, List()) silentResult=SilentResultValue(Test.super.<init>)
normalTypedApply(new Door, List()) silentResult=SilentResultValue(new Door)
normalTypedApply(scala.Predef.int2Integer, List(door.close())) silentResult=SilentResultValue(scala.Predef.int2Integer)
normalTypedApply(door.close, List()) silentResult=SilentResultValue(door.close)
normalTypedApply(scala.Predef.int2Integer(door.close()).toString, List()) silentResult=SilentResultValue(scala.Predef.int2Integer(door.close()).toString)
```

--- a/src/compiler/scala/tools/nsc/typechecker/Infer.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Infer.scala
@@ -1076,7 +1076,7 @@ trait Infer extends Checkable {
      */
     def inferMethodInstance(fn: Tree, undetParams: List[Symbol],
                             args: List[Tree], pt0: Type): List[Symbol] = fn.tpe match {
-      case mt @ MethodType(params0, _) =>
+      case mt @ MethodType(_, _) =>
         try {
           val pt      = if (pt0.typeSymbol == UnitClass) WildcardType else pt0
           val formals = formalTypes(mt.paramTypes, args.length)
@@ -1094,17 +1094,17 @@ trait Infer extends Checkable {
             adjusted.undetParams match {
               case Nil  => Nil
               case xs   =>
-                // #3890
+                // scala/bug#3890
                 val xs1 = treeSubst.typeMap mapOver xs
                 if (xs ne xs1)
                   new TreeSymSubstTraverser(xs, xs1) traverseTrees fn :: args
                 enhanceBounds(adjusted.okParams, adjusted.okArgs, xs1)
                 xs1
             }
-          } else Nil
-        }
-        catch ifNoInstance { msg =>
-          NoMethodInstanceError(fn, args, msg); List()
+          } else undetParams
+        } catch ifNoInstance { msg =>
+          NoMethodInstanceError(fn, args, msg)
+          undetParams
         }
       case x => throw new MatchError(x)
     }

--- a/test/files/neg/leibniz-liskov.check
+++ b/test/files/neg/leibniz-liskov.check
@@ -55,8 +55,8 @@ leibniz-liskov.scala:21: error: type mismatch;
  required: F[U]
   def convert1[T, U](l: List[T])(ev: T =:= U): List[U] = ev.substituteContra(l)
                                                                              ^
-leibniz-liskov.scala:21: error: type mismatch;
- found   : F[T]
+leibniz-liskov.scala:21: error: polymorphic expression cannot be instantiated to expected type;
+ found   : [F[_]]F[T]
  required: List[U]
   def convert1[T, U](l: List[T])(ev: T =:= U): List[U] = ev.substituteContra(l)
                                                                             ^
@@ -72,8 +72,8 @@ leibniz-liskov.scala:22: error: type mismatch;
  required: F[T]
   def convert2[T, U](l: List[U])(ev: T =:= U): List[T] = ev.substituteCo(l)
                                                                          ^
-leibniz-liskov.scala:22: error: type mismatch;
- found   : F[U]
+leibniz-liskov.scala:22: error: polymorphic expression cannot be instantiated to expected type;
+ found   : [F[_]]F[U]
  required: List[T]
   def convert2[T, U](l: List[U])(ev: T =:= U): List[T] = ev.substituteCo(l)
                                                                         ^
@@ -111,8 +111,8 @@ leibniz-liskov.scala:35: error: type mismatch;
  required: F[U]
   def convertConsume2[U, T](c: Consumes[T])(ev: U <:< T): Consumes[U] = ev.substituteCo(c)
                                                                                         ^
-leibniz-liskov.scala:35: error: type mismatch;
- found   : F[T]
+leibniz-liskov.scala:35: error: polymorphic expression cannot be instantiated to expected type;
+ found   : [F[+_]]F[T]
  required: LeibnizLiskov.this.Consumes[U]
     (which expands to)  U => Unit
   def convertConsume2[U, T](c: Consumes[T])(ev: U <:< T): Consumes[U] = ev.substituteCo(c)

--- a/test/files/neg/t12413.check
+++ b/test/files/neg/t12413.check
@@ -1,0 +1,16 @@
+t12413.scala:13: error: inferred type arguments [AnyRef] do not conform to method close's type parameter bounds [Phantom >: AnyRef <: Open]
+  println(door.close.toString())
+               ^
+t12413.scala:14: error: inferred type arguments [AnyRef] do not conform to method close's type parameter bounds [Phantom >: AnyRef <: Open]
+  println(door.close == 0)
+               ^
+t12413.scala:15: error: inferred type arguments [AnyRef] do not conform to method open's type parameter bounds [Phantom >: AnyRef <: Open]
+  println(door.open().toString)
+               ^
+t12413.scala:16: error: inferred type arguments [AnyRef] do not conform to method open's type parameter bounds [Phantom >: AnyRef <: Open]
+  println(door.open().toString())
+               ^
+t12413.scala:17: error: inferred type arguments [AnyRef] do not conform to method open's type parameter bounds [Phantom >: AnyRef <: Open]
+  println(door.open() == 0)
+               ^
+5 errors

--- a/test/files/neg/t12413.scala
+++ b/test/files/neg/t12413.scala
@@ -1,0 +1,18 @@
+class Open
+
+class Door[State] {
+  def close[Phantom >: State <: Open]: Int = 0
+  def open[Phantom >: State <: Open](): Int = 0
+}
+
+class Test {
+  val door = new Door[AnyRef]
+  // the error here happens later (at refchecks)
+  println(door.close.toString)
+  // the errors below happen when typing implicit conversions
+  println(door.close.toString())
+  println(door.close == 0)
+  println(door.open().toString)
+  println(door.open().toString())
+  println(door.open() == 0)
+}

--- a/test/files/neg/t7509.check
+++ b/test/files/neg/t7509.check
@@ -6,7 +6,4 @@ t7509.scala:3: error: type mismatch;
  required: R
   crash(42)
         ^
-t7509.scala:3: error: could not find implicit value for parameter ev: R
-  crash(42)
-       ^
-3 errors
+2 errors

--- a/test/files/neg/t8463.check
+++ b/test/files/neg/t8463.check
@@ -19,9 +19,4 @@ t8463.scala:5: error: type mismatch;
  required: T[Long]
   insertCell(Foo(5))
                  ^
-t8463.scala:5: error: type mismatch;
- found   : Test.Foo[T]
- required: Test.Foo[Test.Cell]
-  insertCell(Foo(5))
-                ^
-4 errors
+3 errors

